File: android/core/observability/src/main/java/com/enterprise/core/observability/UXMetricsTracker.kt
package com.enterprise.core.observability

import com.enterprise.core.domain.interfaces.AnalyticsTracker
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Tracks UX rendering performance metrics (Time to Interactive).
 * Measures time from ViewModel init to first Success state emission.
 */
@Singleton
class UXMetricsTracker @Inject constructor(
    private val analyticsTracker: AnalyticsTracker
) {
    private val startTimes = mutableMapOf<String, Long>()

    /** Call when a screen/component starts loading. */
    fun markStart(tag: String) {
        startTimes[tag] = System.nanoTime()
    }

    /** Call when the first successful state is emitted. Logs TTI metric. */
    fun markInteractive(tag: String) {
        val startTime = startTimes.remove(tag) ?: return
        val elapsedMs = (System.nanoTime() - startTime) / 1_000_000
        analyticsTracker.trackEvent("ux_time_to_interactive", mapOf(
            "screen" to tag,
            "tti_ms" to elapsedMs
        ))
    }
}

File: ios/EnterpriseMobileApp/Sources/Core/Observability/UXMetricsTracker.swift
import Foundation
import CoreDomain

/// Tracks UX performance metrics like Time to Interactive (TTI).
public final class UXMetricsTracker: @unchecked Sendable {
    private let analyticsTracker: AnalyticsTracker
    private var startTimes: [String: CFAbsoluteTime] = [:]
    private let lock = NSLock()

    public init(analyticsTracker: AnalyticsTracker) {
        self.analyticsTracker = analyticsTracker
    }

    /// Call when a screen/feature starts loading.
    public func startMeasuring(_ label: String) {
        lock.lock()
        startTimes[label] = CFAbsoluteTimeGetCurrent()
        lock.unlock()
    }

    /// Call when the screen/feature becomes interactive. Logs the TTI metric.
    public func endMeasuring(_ label: String) {
        lock.lock()
        guard let start = startTimes.removeValue(forKey: label) else {
            lock.unlock()
            return
        }
        lock.unlock()

        let elapsed = CFAbsoluteTimeGetCurrent() - start
        let milliseconds = elapsed * 1000

        analyticsTracker.trackEvent("ux_time_to_interactive", properties: [
            "screen": label,
            "tti_ms": milliseconds
        ])

        // In production: use os_signpost for Instruments integration
        // os_signpost(.end, log: metricsLog, name: "TTI", "%{public}s: %.2fms", label, milliseconds)
    }
}
